File: chitecapi/Controllers/Detalle_AuditoriasController.cs
using chitecapi.Responses;
using DataAccess;
using DataAccess.Models;
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Threading.Tasks;
using System.Web.Http;
using System.Data;
using System.Data.SqlClient;

namespace chitecapi.Controllers
{
    public class Detalle_AuditoriasController : ApiController
    {
        #region Actions

        // GET api/Detalle_Auditorias/ListadoDetalleAuditorias?db=db1
        /// <summary>
        /// Obtiene una lista del detalle de las auditorias.
        /// </summary>
        /// <param name="db" example="db1">The Database ID</param>
        /// <remarks>
        /// Se utiliza para obtener una lista de los articulos en el detalle de las auditoria, generado con un formato JSON:
        /// 
        ///     Example Result:
        ///     {
        ///         "data": [
        ///                   {
        ///                       "id": 1,
        ///                       "id_terminal": "B-4CTTPC2-573D2323",
        ///                       "no_detalleInv": "1",
        ///                       "no_articulo": "25787",
        ///                       "descripcion": "AZUCAR CREMA 1 LB",
        ///                       "id_tipo_ubicacion": 0,
        ///                       "cod_alterno": "G02",
        ///                       "pre_conteo": 2,
        ///                       "cantidad_auditada": 3,
        ///                       "diferencia": 1,
        ///                       "porcentaje_diferencia": 33.33333333,
        ///                       "id_tipo_error": 0,
        ///                       "notas": "",
        ///                       "codigo_barra": "123",
        ///                       "id_usuario_registro": 1,
        ///                       "secuencia_tiro": 1,
        ///                       "alterno1": "",
        ///                       "alterno2": "",
        ///                       "alterno3": "",
        ///                       "cantidad": 0,
        ///                       "costo": 0,
        ///                       "costo_total": 0,
        ///                       "precio": 0,
        ///                       "precio_total": 0,
        ///                       "id_ubicacion": 0,
        ///                       "fecha_registro": "0000-00-00 00:00:00",
        ///                       "fecha_modificacion": "0000-00-00 00:00:00",
        ///                       "id_usuario_modificacion": 0,
        ///                       "id_auditor": 0,
        ///                       "id_tipo_auditoria": 0,
        ///                       "estado": 0,
        ///                       "created_at": "0000-00-00 00:00:00",
        ///                       "updated_at": "0000-00-00 00:00:00",
        ///                       "es_cantidad_igual_tiros": 0
        ///                    }
        ///                 ],
        ///         "error": 0,
        ///         "error_type": 0,
        ///         "error_message": 0
        ///     }
        /// </remarks>
        /// <response code="401">Unauthorized. Error en la configuracion de la base de datos</response>              
        /// <response code="200">OK. Devuelve el objeto solicitado.</response>        
        /// <response code="404">NotFound. No se ha encontrado el objeto solicitado.</response>
        [HttpGet]
        public IHttpActionResult ListadoDetalleAuditorias([FromUri] string db = "db1")
        {

            string connetionString;
            SqlConnection conection;
            string dbconection = "db1";

            if (!db.Equals(""))
            {
                dbconection = db;
            }
            connetionString = ConfigurationManager.ConnectionStrings[dbconection].ConnectionString;
            conection = new SqlConnection(connetionString);

            conection.Open();


            var sql = ConfigurationManager.AppSettings["getdetalleauditorias"];

            if (sql == null)
            {
                sql = " SELECT *, case  when d.cantidad=d.no_detalleInv then 1 else 0 end as es_cantidad_igual_tiros FROM detalle_auditorias d ";
            }



            SqlDataAdapter dataAdapter = new SqlDataAdapter(sql, conection);


            DataTable table = new DataTable();
            dataAdapter.Fill(table);

            conection.Close();
            Dictionary<string, object> jsonvalues = new Dictionary<string, object>();
            jsonvalues.Add("data", table);
            jsonvalues.Add("error", 0);
            jsonvalues.Add("error_type", 0);
            jsonvalues.Add("error_message", 0);


            return Json(jsonvalues);
        }




        [HttpPost]
        public async Task<IHttpActionResult> Guardar([FromBody] List<DetalleAuditoria> partialDetallesAuditoria, string db)
        {
            if (string.IsNullOrEmpty(db))
            {
                db = $"{ConfigurationManager.AppSettings["default_db"]}";
            }

            if (ConfigurationManager.ConnectionStrings[db] == null)
            {
                return new CustomJsonActionResult(
                    System.Net.HttpStatusCode.NotFound,
                    new JsonErrorResponse(1, 400, $"La base de datos {db} no existe."));
            }

            var dbAccess = new SqlServerDataAccess(
                ConfigurationManager.ConnectionStrings[db].ConnectionString, ConfigurationManager.ConnectionStrings[db].ProviderName
                );

            foreach (var partialDetalleAuditoria in partialDetallesAuditoria)
            {
                if (partialDetalleAuditoria?.IsNotInitialized == true)
                {
                    return new CustomJsonActionResult(
                        System.Net.HttpStatusCode.NotFound,
                        new JsonErrorResponse(1, 400, "Faltan parámetros"));
                }

                try
                {
                    var articulo = await GetArticulo(partialDetalleAuditoria.no_articulo, dbAccess);

                    if (articulo == null)
                    {
                        return new CustomJsonActionResult(
                            System.Net.HttpStatusCode.NotFound,
                            new JsonErrorResponse(1, 1, $"El artículo con el código  {partialDetalleAuditoria.no_articulo} no existe en la base de datos."));
                    }

                    var fullDetalleAuditoria =
                        SetFullDetalleAuditoria(partialDetalleAuditoria, articulo);

                    var sql = ConfigurationManager.AppSettings["crear_detalle_auditoria"];

                    var parameters = new
                    {
                        id_terminal = fullDetalleAuditoria.id_terminal,
                        no_detalleInv = fullDetalleAuditoria.no_detalleInv,
                        no_articulo = fullDetalleAuditoria.no_articulo,
                        descripcion = fullDetalleAuditoria.descripcion,
                        id_tipo_ubicacion = fullDetalleAuditoria.id_tipo_ubicacion,
                        cod_alterno = fullDetalleAuditoria.cod_alterno,
                        pre_conteo = fullDetalleAuditoria.pre_conteo,
                        cantidad_auditada = fullDetalleAuditoria.cantidad_auditada,
                        diferencia = fullDetalleAuditoria.diferencia,
                        porcentaje_diferencia = fullDetalleAuditoria.porcentaje_diferencia,
                        id_tipo_error = fullDetalleAuditoria.id_tipo_error,
                        notas = fullDetalleAuditoria.notas,
                        codigo_barra = fullDetalleAuditoria.codigo_barra,
                        id_usuario_registro = fullDetalleAuditoria.id_usuario_registro,
                        secuencia_tiro = fullDetalleAuditoria.secuencia_tiro,
                        alterno1 = fullDetalleAuditoria.alterno1,
                        alterno2 = fullDetalleAuditoria.alterno2,
                        alterno3 = fullDetalleAuditoria.alterno3,
                        cantidad = fullDetalleAuditoria.cantidad,
                        costo = fullDetalleAuditoria.costo,
                        costo_total = fullDetalleAuditoria.costo_total,
                        precio = fullDetalleAuditoria.precio,
                        precio_total = fullDetalleAuditoria.precio_total,
                        id_ubicacion = fullDetalleAuditoria.id_ubicacion,
                        fecha_registro = fullDetalleAuditoria.fecha_registro,
                        fecha_modificacion = fullDetalleAuditoria.fecha_modificacion,
                        id_usuario_modificacion = fullDetalleAuditoria.id_usuario_modificacion,
                        id_auditor = fullDetalleAuditoria.id_auditor,
                        id_tipo_auditoria = fullDetalleAuditoria.id_tipo_auditoria,
                        estado = 1,
                        created_at = DateTime.Now,
                        updated_at = DateTime.Now
                    };

                    await dbAccess.SaveDataAsync(sql, parameters);

                }
                catch (Exception exception)
                {
                    return new CustomJsonActionResult(
                        System.Net.HttpStatusCode.NotFound,
                        new JsonErrorResponse(1, 1, exception.GetBaseException().Message));
                }
            }

            return new CustomJsonActionResult(
                    System.Net.HttpStatusCode.OK,
                    new JsonErrorResponse(0, 0, "0"));
        }

        #endregion

        #region Helper methods

        private DetalleAuditoria SetFullDetalleAuditoria(DetalleAuditoria partialDetalleAuditoria, Articulo articulo)
        {
            float.TryParse(articulo.costo, out var articuloCosto);
            float.TryParse(articulo.precio, out var articuloPrecio);
            float.TryParse(partialDetalleAuditoria.cantidad, out var articuloCantidad);

            partialDetalleAuditoria.costo_total = (articuloCosto * articuloCantidad).ToString();
            partialDetalleAuditoria.precio_total = (articuloPrecio * articuloCantidad).ToString();
            partialDetalleAuditoria.alterno1 = articulo.alterno1;
            partialDetalleAuditoria.alterno2 = articulo.alterno2;
            partialDetalleAuditoria.alterno3 = articulo.alterno3;
            partialDetalleAuditoria.costo = articulo.costo;
            partialDetalleAuditoria.precio = articulo.precio;

            return partialDetalleAuditoria;
        }

        private async Task<Articulo> GetArticulo(string articuloNumber, SqlServerDataAccess dbAccess)
        {
            var sql = ConfigurationManager.AppSettings["buscar_articulo_por_numero"];

            return
                await dbAccess.LoadFirstOrDefaultAsync<Articulo, dynamic>(
                    sql, new { no_articulo = articuloNumber });
        }

        #endregion
    }
}

File: Legacy/PHP Old Api/Controllers/Detalle_auditorias_controller.php
<?php
namespace App\Http\Controllers;

use Illuminate\Http\Request;
use stdClass;
use Symfony\Component\HttpKernel\Exception\HttpException;
use DB;

class Detalle_auditorias_controller extends Controller
{
    public function index(Request $request) 
    {
        $json = new stdClass(); 
        $es_request_valido = validar_request(["DB"], $json, $request);
        if ($es_request_valido) 
        {
            $DB = ($request->input('DB') != NULL) ? $request->input('DB') : '';
            $existe_base_datos = DB::select('SELECT 1
             FROM INFORMATION_SCHEMA.SCHEMATA
             WHERE SCHEMA_NAME = ?
             LIMIT 1', [
                $DB
            ]);
            $existe_base_datos = (count($existe_base_datos) > 0);
            if ($existe_base_datos) 
            {
                //Validamos el token
                //if (validar_token($request->input('token'), $json, $DB)) 
                //{
                    $detalle_auditorias = DB::select("SELECT * FROM {$DB}.detalle_auditorias");
                    foreach ($detalle_auditorias as $ind => $detalle_auditoria) 
                    {
                        $es_cantidad_igual_tiros = ($detalle_auditoria->cantidad == $detalle_auditoria->no_detalleInv) ? 1 : 0;
                        $detalle_auditorias[$ind]->es_cantidad_igual_tiros = $es_cantidad_igual_tiros;
                    }

                    $json->data = $detalle_auditorias;
                    $json->error = 0;
                    $json->error_type = 0;
                    $json->error_message = 0;    
                //}
            }
            else 
            {
                $json->error = 1;
                $json->error_type = 400;
                $json->error_message = "La base de datos: \"{$DB}\" no existe";
            } 
        }

        if ($json->error_type != 400) 
        {
            return response()->json(
                $json,
                200
            );
        } else 
        {
            throw new HttpException(400, $json->error_message);
        }
    }

    public function guardar(Request $request) 
    {
        $fecha_hoy = date('Y-m-d H:i:s');
        $request = json_decode(file_get_contents('php://input'), true);
        $json = new stdClass();
        if ($request != NULL && isset($_GET['DB']))
        {
            $DB = $_GET['DB'];
            $existe_base_datos = DB::select('SELECT 1
             FROM INFORMATION_SCHEMA.SCHEMATA
             WHERE SCHEMA_NAME = ?
             LIMIT 1', [
                $DB
            ]);
            $existe_base_datos = (count($existe_base_datos) > 0);
            if ($existe_base_datos) 
            {
                $datos_detalle_auditorias = $request;
                foreach ($datos_detalle_auditorias as $datos_detalle_auditoria)
                {
                    $articulo = DB::select("SELECT codigo_barra, alterno1, alterno2, alterno3, descripcion,
                    costo, precio
                    FROM {$DB}.articulos
                    WHERE no_articulo = ? LIMIT 1", [
                        $datos_detalle_auditoria['no_articulo']
                    ]);
                   if (count($articulo) > 0)
                   {
                        $articulo = $articulo[0];

                        $costo_total_detalle = (floatval($articulo->costo) * floatval($datos_detalle_auditoria['cantidad']));
                        $precio_total_detalle = (floatval($articulo->precio) * floatval($datos_detalle_auditoria['cantidad']));
                        DB::insert("INSERT INTO {$DB}.detalle_auditorias (id_terminal, no_detalleInv, no_articulo, descripcion, id_tipo_ubicacion, cod_alterno, pre_conteo, 
                            cantidad_auditada, diferencia, porcentaje_diferencia, id_tipo_error, notas, codigo_barra, id_usuario_registro, secuencia_tiro,
                            alterno1, alterno2, alterno3, cantidad, costo, costo_total, precio, precio_total, id_ubicacion, fecha_registro,
                            fecha_modificacion, id_usuario_modificacion, id_auditor, id_tipo_auditoria, estado, created_at, updated_at)
                            VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,
                            ?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,1,?,?)", [
                            $datos_detalle_auditoria['id_terminal'],
                            $datos_detalle_auditoria['no_detalleInv'],
                            $datos_detalle_auditoria['no_articulo'],
                            $datos_detalle_auditoria['descripcion'],
                            $datos_detalle_auditoria['id_tipo_ubicacion'],
                            $datos_detalle_auditoria['cod_alterno'],
                            $datos_detalle_auditoria['pre_conteo'],
                            $datos_detalle_auditoria['cantidad_auditada'],
                            $datos_detalle_auditoria['diferencia'],
                            $datos_detalle_auditoria['porcentaje_diferencia'],
                            $datos_detalle_auditoria['id_tipo_error'],
                            $datos_detalle_auditoria['notas'],
                            $datos_detalle_auditoria['codigo_barra'],
                            $datos_detalle_auditoria['id_usuario_registro'],
                            $datos_detalle_auditoria['secuencia_tiro'],
                            $articulo->alterno1,
                            $articulo->alterno2,
                            $articulo->alterno3,
                            $datos_detalle_auditoria['cantidad'],
                            $articulo->costo,
                            $costo_total_detalle,
                            $articulo->precio,
                            $precio_total_detalle,
                            $datos_detalle_auditoria['id_ubicacion'],
                            $datos_detalle_auditoria['fecha_registro'],
                            $datos_detalle_auditoria['fecha_modificacion'],
                            $datos_detalle_auditoria['id_usuario_modificacion'],
                            $datos_detalle_auditoria['id_auditor'],
                            $datos_detalle_auditoria['id_tipo_auditoria'],
                            $fecha_hoy,
                            $fecha_hoy
                        ]);
                    }
                    else
                    {
                        $json->error = 1;
                        $json->error_type = 1;
                        $json->error_message = "El artículo con el código: \"{$datos_detalle_auditoria['no_articulo']}\" no existe en la base de datos";

                        if ($json->error_type != 400)
                        {
                            return response()->json(
                                $json,
                                200
                            );
                        } else
                        {
                            throw new HttpException(400, $json->error_message);
                        }
                        exit();
                    }
                }

                $json->error = 0;
                $json->error_type = 0;
                $json->error_message = 0;
            }
            else 
            {
                $json->error = 1;
                $json->error_type = 400;
                $json->error_message = "La base de datos: \"{$DB}\" no existe";
            } 
        }
        else
        {
            $json->error = 1;
            $json->error_type = 400;
            $json->error_message = 'Faltan parámetros';
        }

        if ($json->error_type != 400)
        {
            return response()->json(
                $json,
                200
            );
        } else
        {
            throw new HttpException(400, $json->error_message);
        }
    }
}